### PR TITLE
feat: generate .meta/prefixes.txt for Maven repository versions

### LINKED ISSUE #399: Generate .meta/prefixes.txt for Maven repository versions
## Problem Statement

Maven Central publishes a `.meta/prefixes.txt` file that lists all group ID path prefixes available in the repository. Repository managers like Artifactory and Nexus use this file to quickly determine which repositories might contain a given artifact, enabling efficient routing without making individual requests to every configured remote repository.

Pulp Maven repositories do not generate this file. This limits discoverability when Pulp repositories are used as upstream sources for Artifactory or Nexus.

## File Format

The file follows the `repository-prefixes/2.0` format used by Maven Central:

```
## repository-prefixes/2.0
/com/fasterxml
/com/jayway
/commons-fileupload
/io/smallrye
/net/minidev
/org/springframework
```

- Header line: `## repository-prefixes/2.0`
- One line per unique prefix, sorted alphabetically
- Each line starts with `/`
- The prefix is the first two path segments of the group ID (e.g. `com.fasterxml.jackson.core` → `/com/fasterxml`)
- Single-segment group IDs use the full name (e.g. `commons-fileupload` → `/commons-fileupload`)

## Requirements

- The file should be generated as part of `finalize_new_version` whenever Maven artifacts are added or removed
- It should be served by the content app at `<base_path>/.meta/prefixes.txt`
- The file should reflect the content of the current repository version

## Reference

Maven Central's prefixes.txt: https://repo.maven.apache.org/maven2/.meta/prefixes.txt (23,000+ entries)

Jira: PULP-2003

### PR DESCRIPTION
**Pull request by CryptoRodeo**
Automatically generate a `.meta/prefixes.txt` file listing top-level group ID prefixes in Maven repositories. 

This file allows Maven clients to efficiently determine which repositories to query.

- Add `_compute_prefixes` and `save_prefixes_txt` helpers to models
- Generate prefixes.txt during `_generate_metadata` and `repair_metadata`
- Only regenerate when prefixes actually change
- Add functional tests covering prefix generation, updates, removal,
  and content app download

    Closes: https://github.com/pulp/pulp_maven/issues/399
### 📜 Checklist

- [x] Commits are cleanly separated with meaningful messages (simple features and bug fixes should be [squashed](https://pulpproject.org/pulpcore/docs/dev/guides/git/#rebasing-and-squashing) to one commit)
- [x] A [changelog entry](https://pulpproject.org/pulpcore/docs/dev/guides/git/#changelog-update) or entries has been added for any significant changes
- [x] Follows the [Pulp policy on AI Usage](https://pulpproject.org/help/more/governance/ai_policy/)
- [x] (For new features) - User documentation and test coverage has been added

See: [Pull Request Walkthrough](https://pulpproject.org/pulpcore/docs/dev/guides/pull-request-walkthrough/)

### CODE REVIEW

File: pulp_maven/app/models.py [hunk 1]
```diff
@@ -384,5 +387,54 @@ def _generate_metadata(self, new_version):
...
+        current_prefixes = {_compute_prefix(gid) for gid in all_group_ids}
+
+        unaffected_group_ids = set(
+            MavenArtifact.objects.filter(pk__in=new_version.content)
+            .exclude(pairs_q)
+            .values_list("group_id", flat=True)
+            .distinct()
+        )
+
+        unaffected_prefixes = {_compute_prefix(gid) for gid in unaffected_group_ids}
```
Comment (dkliban) [inline pulp_maven/app/models.py:420]: When this runs for the first time on a repository that has never had `prefixes.txt`, the file won't be generated unless the change happens to introduce a new prefix.

Trace through the scenario: existing repo with many artifacts, someone adds an artifact under an already-existing prefix (e.g. another `com.example.*` when `com.example` artifacts already exist):

1. `affected_prefixes` = `{/com/example}`
2. `unaffected_prefixes` also contains `/com/example` (other artifacts share that prefix)
3. `prefix_set_changed = bool({/com/example} - {/com/example, ...})` → `False`
4. No `prefixes.txt` is generated

The fix would be to also check whether `prefixes.txt` already exists in the version. If it's missing, generate it with all current prefixes regardless of whether the prefix set changed. Something like:



The `repair_metadata` path handles this correctly since it unconditionally rebuilds, but the normal `_generate_metadata` path has this gap.

Comment (CryptoRodeo) [reply pulp_maven/app/models.py:420]: Good catch, let me fix this...

Comment (CryptoRodeo) [reply pulp_maven/app/models.py:420]: I uploaded a fix, now creating a functional test

File: pulp_maven/app/tasks/__init__.py [hunk 2]
```diff
@@ -32,6 +32,73 @@
...
+
+    The prefix is the first two slash-separated segments of the group path,
+    or the full path if the group ID has fewer than two segments.
+
+    Examples:
+        com.fasterxml.jackson.core -> /com/fasterxml
+        com.springframework.boot   -> /org/springframework
+        commons-fileupload         -> /commons/file-upload
+    """
+    group_path = group_id.replace(".", "/")
```
Comment (dkliban) [inline pulp_maven/app/tasks/__init__.py:49]: The second and third docstring examples are incorrect:

- `com.springframework.boot` would produce `/com/springframework`, not `/org/springframework`
- `commons-fileupload` has no dots, so `replace(".", "/")` is a no-op — it produces `/commons-fileupload` (single segment), not `/commons/file-upload`

Suggested fix:
```python
Examples:
    com.fasterxml.jackson.core -> /com/fasterxml
    com.springframework.boot   -> /com/springframework
    commons-fileupload         -> /commons-fileupload
```

File: pulp_maven/app/tasks/__init__.py [hunk 3]
```diff
@@ -32,6 +32,73 @@
     "maven-metadata.xml.sha1",
     "maven-metadata.xml.sha256",
 ]
```
Comment (dkliban) [inline pulp_maven/app/tasks/__init__.py:34]: Pull-through interaction concern: `MavenRemote.get_remote_artifact_content_type` returns `MavenArtifact` for `.meta/prefixes.txt` paths since it doesn't match any of the exclusion patterns (which only cover `maven-metadata.xml` and its checksums).

In a pull-through repository, a client requesting `.meta/prefixes.txt` would cause the handler to try fetching it from the upstream remote and saving it as a `MavenArtifact` — which could either fail (404 from upstream) or, worse, succeed and store the upstream's `prefixes.txt` as a `MavenArtifact` content unit that conflicts with the locally-generated `MavenMetadata` version.

Consider adding `.meta/prefixes.txt` to the exclusion list in `get_remote_artifact_content_type` (returning `None`) so the pull-through handler either streams it or skips it cleanly.

File: pulp_maven/tests/functional/api/test_prefixes_txt.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+        ).task
+    )
+
+    prefixes_url = urljoin(base_url, ".meta/prefixes.txt")
+    downloaded = download_file(prefixes_url)
+    assert downloaded.response_obj.status == 200
+
+    text = downloaded.body.decode("utf-8")
+    assert text.startswith("## repository-prefixes/2.0\n")
+    assert f"/com/{uid}" in text
```
Comment (dkliban) [inline pulp_maven/tests/functional/api/test_prefixes_txt.py:257]: Missing test for the "remove all artifacts" edge case. When every artifact is removed from a repository, `prefixes.txt` should also be removed (not left stale). The code handles this correctly (`current_prefixes` is empty so no new file is created, and the old one is removed), but a test would guard against regressions.

Something like: add two artifacts with different prefixes, verify `prefixes.txt` exists, remove both, verify `prefixes.txt` is gone (either 404 on download or metadata count is 0).

File: pulp_maven/app/tasks/__init__.py (hunk 2 above)
Comment (CryptoRodeo) [reply pulp_maven/app/tasks/__init__.py:49]: updated here: https://github.com/pulp/pulp_maven/pull/409/commits/b8472046e4d9d422b2d8192bebafbffd7a17d5e6

File: pulp_maven/app/tasks/__init__.py (hunk 3 above)
Comment (CryptoRodeo) [reply pulp_maven/app/tasks/__init__.py:34]: Updated here: https://github.com/pulp/pulp_maven/pull/409/commits/b8472046e4d9d422b2d8192bebafbffd7a17d5e6

File: pulp_maven/tests/functional/api/test_prefixes_txt.py (hunk 4 above)
Comment (CryptoRodeo) [reply pulp_maven/tests/functional/api/test_prefixes_txt.py:257]: Added here: https://github.com/pulp/pulp_maven/pull/409/commits/fb2dbb17bee16e156777bcdfe2bffc10b5fc0114

**Comment by CryptoRodeo** [comment]
Commits squashed

**Review by dkliban — APPROVED** [review]
